polytracker/tracing.py

```diff
@@ -1077,6 +1095,51 @@ def taint_forest(self) -> TaintForest:
         """The taint forest associated with this trace."""
         raise NotImplementedError()
 
+    @abstractmethod
+    def file_offset(self, node: TaintForestNode) -> ByteOffset:
+        """The file offset associated with a taint forest node"""
+        raise NotImplementedError()
+
+    def inputs_affecting_control_flow(self) -> Taints:
+        """Returns the set of byte offsets that affected control flow"""
+        return self.taints((node for node in self.taint_forest.nodes() if node.affected_control_flow))
+
+    def taints(self, labels: Iterable[TaintForestNode]) -> Taints:
+        # reverse the labels to reduce the likelihood of reproducing work
+        history: Set[TaintForestNode] = set(labels)
+        node_stack: List[TaintForestNode] = sorted(list(set(history)), reverse=True)
+        taints: Set[ByteOffset] = set()
+        if len(node_stack) < 10:
+            labels_str = ", ".join(map(str, node_stack))
+        else:
+            labels_str = f"{len(node_stack)} labels"
+        with tqdm(
+                desc=f"finding canonical taints for {labels_str}",
+                leave=False,
+                delay=5.0,
+                bar_format="{l_bar}{bar}| [{elapsed}<{remaining}, {rate_fmt}{postfix}]'",
+                total=sum(node.label for node in node_stack),
+        ) as t:
+            while node_stack:
+                node = node_stack.pop()
+                t.update(node.label)
+                if node.parent_one is None:
+                    assert node.parent_two is None
+                    taints.add(self.file_offset(node))
+                else:
+                    parent1, parent2 = node.parent_one, node.parent_two
+                    assert parent1 is not None and parent2 is not None
```

Comment:
Why is it required for a node to have two parents? Is this because node labels are reused if it is only involved in an unary expression? Maybe a minor comment here about the reasoning?

Reply:
That is correct.
```suggestion
                    // a node will always have either zero or two parents.
                    // labels that are reused will reuse their associated nodes.
                    // all other nodes are unions.
                    assert parent1 is not None and parent2 is not None
```